core/src/main/scala/org/apache/spark/h2o/converters/PrimitiveRDDConverter.scala

```diff
@@ -57,25 +65,29 @@ private[converters] object PrimitiveRDDConverter extends Logging with ConverterU
   def perPrimitiveRDDPartition[T]() // extra arguments for this transformation
                                  (keyName: String, vecTypes: Array[Byte], uploadPlan: Option[immutable.Map[Int, NodeDesc]]) // general arguments
                                  (context: TaskContext, it: Iterator[T]): (Int, Long) = { // arguments and return types needed for spark's runJob input
+
+    val asArr = it.toArray[Any] // need to buffer the iterator in order to get its length
     val con = ConverterUtils.getWriteConverterContext(uploadPlan, context.partitionId())
+    con.createChunks(keyName, vecTypes, context.partitionId(), asArr.length)
 
-    con.createChunks(keyName, vecTypes, context.partitionId())
-    // try to wait for reply to ensure we can continue with sending
-    it.foreach { r =>
-      r match {
-        case n: Number => con.put(0, n)
-        case n: Boolean => con.put(0, n)
-        case n: String => con.put(0, n)
-        case n: java.sql.Timestamp => con.put(0, n)
-        case _ => con.putNA(0)
-      }
-      con.increaseRowCounter()
+    asArr.foreach {
+      case n: Boolean => con.put(0, n)
+      case n: Byte => con.put(0, n)
+      case n: Char => con.put(0, n)
+      case n: Short => con.put(0, n)
+      case n: Int => con.put(0, n)
+      case n: Long => con.put(0, n)
+      case n: Float => con.put(0, n)
+      case n: Double => con.put(0, n)
+      case n: String => con.put(0, n)
+      case n: java.sql.Timestamp => con.put(0, n)
+      case _ => con.putNA(0)
     }
```

Comment:
This chunk of code was already somewhere above.
Why not extract a method/function for it?